### PR TITLE
feat: add "join server" modal

### LINKED ISSUE #102: Make create server button multi-action
**Suggested by eksdi (01FBYHDMTPM3APK7X9ARQ5S5R3) in Suggestions on Revolt server**

Current "Create Server" button should now show a modal with the following options

- Join server
- Create server

### LINKED ISSUE #436: add "join server" form
we could either make the plus button on the server list a choice between creating a server and using an invite (preferable but might need some design work), or add a separate button (a bit clunky)

### PR DESCRIPTION
**Pull request by Rexogamer**
replacing #468, which was closed due to a branch move

fixes #436
fixes #102

### CODE REVIEW

File: src/context/intermediate/modals/Input.tsx [hunk 1]
```diff
@@ -139,6 +140,33 @@ export function SpecialInputModal(props: SpecialProps) {
...
+                        const server = await client.fetchInvite(code);
+                        if (typeof server === "undefined")
+                            console.log("Something went wrong.");
+
+                        if (client.servers.get(server.server_id)) {
+                            history.push(
+                                `/server/${server.server_id}/channel/${server.channel_id}`,
+                            );
+                        }
+                        await client.joinInvite(code);
```
Comment (insertish) [inline src/context/intermediate/modals/Input.tsx:165]: Should probably open the server after joining, although for this the logic from the invite page and embed should be combined and also used here.

**Review by insertish — CHANGES_REQUESTED** [review]

File: src/context/intermediate/modals/Input.tsx [hunk 2]
```diff
@@ -139,6 +140,33 @@ export function SpecialInputModal(props: SpecialProps) {
...
+                    callback={async (rawCode) => {
+                        // if the user provides an invite link, get rid of the url and just pass the invite code
+                        // prettier-ignore
+                        const regex = new RegExp(
+                            "http(s?):\/\/(app|nightly|rvlt|localhost).(revolt.chat|gg\/|\d{3,5})(\/invite\/)?", // localhost doesn't quite work yet
+                        );
+                        const code = rawCode.replace(regex, "");
+                        const server = await client.fetchInvite(code);
+                        if (typeof server === "undefined")
+                            console.log("Something went wrong.");
```
[the remark this answers is not in the document]
Comment (Rexogamer) [reply src/context/intermediate/modals/Input.tsx:158]: I did attempt this, will try again

[the remark this answers is not in the document]
Comment (insertish) [reply src/context/intermediate/modals/Input.tsx:158]: (any errors thrown here should be caught by the modal and displayed)

**Comment by insertish** [comment]
Closing due to rewrite, https://github.com/revoltchat/frontend.
We should make this from the get-go with the new client.

**Pull request closed without merging**